lib/galaxy/webapps/galaxy/api/jobs.py

```diff
@@ -230,19 +231,15 @@ def __get_job( self, trans, id ):
             decoded_job_id = self.decode_id( id )
         except Exception:
             raise exceptions.MalformedId()
-        query = trans.sa_session.query( trans.app.model.Job )
-        if trans.user_is_admin():
-            query = query.filter(
-                trans.app.model.Job.id == decoded_job_id
-            )
-        else:
-            query = query.filter(
-                trans.app.model.Job.user == trans.user,
-                trans.app.model.Job.id == decoded_job_id
-            )
-        job = query.first()
+        job = trans.sa_session.query( trans.app.model.Job ).filter( trans.app.model.Job.id == decoded_job_id ).first()
         if job is None:
             raise exceptions.ObjectNotFound()
+        if not trans.user_is_admin() and job.user != trans.user:
+            history = self.history_manager.get_current( trans )
+            if history is None:
+                raise exceptions.ObjectNotFound()
+            if len( [ hda.id for hda in history.datasets if hda.creating_job.id == job.id ] ) == 0:
```

Comment:
Thanks @natefoo. We do have the dataset id and used it in the past but after #739 we started using the creating job id instead. This seemed to be reasonable at the time. It also avoided separate build_for_rerun api endpoints (1. dataset, 2. jobs). However the permission handling is not straightforward anymore and causes issues with imported datasets.